### PR TITLE
Export the SkClipOp argument in Canvas.clipRect

### PR DESCRIPTION
**Pull request by jason-simmons**

### CODE REVIEW

**Comment by jason-simmons** [comment]
@chinmaygarde @HansMuller

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -1461,6 +1461,15 @@ enum PointMode {
   polygon,
 }
 
+/// Defines how a new clip region should be merged with the existing clip
+/// region.
+///
+/// Used by [Canvas.clipRect].
+enum ClipOp {
+  difference,
+  intersect,
```
Comment (Hixie) [inline lib/ui/painting.dart]: the values must be documented (see other enums for style)

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -1692,14 +1701,15 @@ class Canvas extends NativeFieldWrapperClass2 {
   /// intersect with the clip boundary, this can result in incorrect blending at
   /// the clip boundary. See [saveLayer] for a discussion of how to address
   /// that.
-  void clipRect(Rect rect) {
+  void clipRect(Rect rect, {ClipOp clipOp: ClipOp.intersect}) {
```
Comment (Hixie) [inline lib/ui/painting.dart]: nit: space after the `{` and before the `}`

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -1692,14 +1701,15 @@ class Canvas extends NativeFieldWrapperClass2 {
...
     assert(_rectIsValid(rect));
-    _clipRect(rect.left, rect.top, rect.right, rect.bottom);
```
Comment (Hixie) [inline lib/ui/painting.dart]: assert clipOp isn't null

File: lib/ui/painting.dart [hunk 4]
```diff
@@ -1692,14 +1701,15 @@ class Canvas extends NativeFieldWrapperClass2 {
   /// intersect with the clip boundary, this can result in incorrect blending at
   /// the clip boundary. See [saveLayer] for a discussion of how to address
   /// that.
-  void clipRect(Rect rect) {
```
Comment (Hixie) [inline lib/ui/painting.dart]: document the argument

**Comment by Hixie** [comment]
What is the motivation for this patch?

**Comment by jason-simmons** [comment]
@HansMuller requested this for a framework feature he was working on

@HansMuller is this still useful?  Or does Canvas.clipPath provide what you need?

**Comment by jason-simmons** [comment]
Comments addressed - PTAL

**Comment by Hixie** [comment]
LGTM

If you want extra bonus points, you could make diagrams for the enum docs the same way that we have for TileMode. :-)

**Comment by jason-simmons** [comment]
Fixes https://github.com/flutter/flutter/issues/12617

**Pull request merged**